modules/nlu/src/backend/engine2.ts

```diff
@@ -1085,23 +1092,23 @@ const predict = {
 
     return _.merge(_.cloneDeep(input), { intent_predictions: { ambiguous } })
   },
-  ExtractSlots: async (input: PredictStepOutput, predictors: Predictors) => {
+  ExtractSlots: async (input: PredictStep) => {
     const intent =
       !input.intent_predictions.ambiguous && input.intents.find(i => i.name === input.intent_predictions.elected.name)
     if (intent && intent.slot_definitions.length > 0) {
       // TODO try to extract for each intent predictions and then rank this in the election step
-      const slots = await predictors.slot_tagger.extract(input.utterance, intent)
+      const slots = await input.predictors.slot_tagger.extract(input.utterance, intent)
       slots.forEach(({ slot, start, end }) => {
         input.utterance.tagSlot(slot, start, end)
       })
     }
 
-    return input
+    return { ...input }
```

Comment:
Not sure to see the reason for that change, since the end result is the same

Reply:
Result is not the same, we create a copy. (almost) every step of the training and the predict pipeline are immutable.